### PR TITLE
`ModelComparisonSimulator`: handle different outputs from individual simulators

### PR DESCRIPTION
**Pull request by Kucharssim**
fixes https://github.com/bayesflow-org/bayesflow/issues/441

As per https://github.com/bayesflow-org/bayesflow/issues/441#issuecomment-2830285318, this PR implements

> "smarter" concatenation function (for example one that pads with nan if a parameter is not available for a given model). 

However, by default the simulator will just drop (with an info warning) keys that are not common for all simulators, since in most situations we would not need those outputs in the first place.

### CODE REVIEW

**Comment by vpratz** [comment]
Thanks for the PR, I skimmed it and like the idea behind the changes. I'll try to conduct a proper review some time this week.

File: bayesflow/simulators/model_comparison_simulator.py [hunk 1]
```diff
@@ -38,8 +40,14 @@ def __init__(
...
-            If True, samples in a batch are drawn from different models. If False, the entire batch
-            is drawn from a single model chosen according to the model probabilities. Default is True.
+            Whether to draw samples in a batch from different models.
+            - If True (default), each sample in a batch may come from a different model.
+            - If False, the entire batch is drawn from a single model, selected according to model probabilities.
+        key_conflicts : {"drop"} | float, optional
+            Policy for handling keys that are missing in the output of some models, when using mixed batches.
+            - "drop" (default): Drop conflicting keys from the batch output.
+            - float: Fill missing keys with the specified value.
+            - If neither "drop" nor a float is given, an error is raised when key conflicts are detected.
```
Comment (LarsKue) [inline bayesflow/simulators/model_comparison_simulator.py:50]: I would make this into `key_conflicts="error"`.

File: bayesflow/simulators/model_comparison_simulator.py [hunk 2]
```diff
@@ -38,8 +40,14 @@ def __init__(
             A sequence of logits corresponding to model probabilities. Mutually exclusive with `p`.
             If neither `p` nor `logits` is provided, defaults to uniform logits.
         use_mixed_batches : bool, optional
-            If True, samples in a batch are drawn from different models. If False, the entire batch
-            is drawn from a single model chosen according to the model probabilities. Default is True.
+            Whether to draw samples in a batch from different models.
+            - If True (default), each sample in a batch may come from a different model.
+            - If False, the entire batch is drawn from a single model, selected according to model probabilities.
+        key_conflicts : {"drop"} | float, optional
```
Comment (LarsKue) [inline bayesflow/simulators/model_comparison_simulator.py:46]: I get the idea behind `key_conflicts=float`, but I think it is confusing from the outside. Instead, I would choose something like `key_conflicts="replace"` or `"fill"` and have an additional argument `fill_value=None`.

**Review by LarsKue — COMMENTED** [review]
Looks good from my side. See individual comments.

Can we also add tests for the (few) missed edge-case lines?

File: bayesflow/simulators/model_comparison_simulator.py (hunk 2 above)
Comment (Kucharssim) [reply bayesflow/simulators/model_comparison_simulator.py]: Yes, that improves clarity of the interface. Thanks!

File: bayesflow/simulators/model_comparison_simulator.py (hunk 1 above)
Comment (Kucharssim) [reply bayesflow/simulators/model_comparison_simulator.py]: Done

File: bayesflow/simulators/model_comparison_simulator.py [hunk 3]
```diff
@@ -118,3 +133,58 @@ def sample(self, batch_shape: Shape, **kwargs) -> dict[str, np.ndarray]:
...
+                    shape[0] = batch_sizes[i]
+                    sim[missing_key] = np.full(shape=shape, fill_value=self.fill_value)
+            return sims
+        elif self.key_conflicts == "error":
+            raise ValueError("Key conflicts are found in simulator outputs, cannot combine them into one batch.")
+
+    def _determine_key_conflicts(self, sims):
+        # determine only once
+        if self._keys is not None:
+            return self._keys
```
Comment (vpratz) [inline bayesflow/simulators/model_comparison_simulator.py:168]: Can this return "wrong" results if some simulators had `n=0` in line 120 when the function first runs, and `n>0` later on? Is this something we want to safeguard against?

Comment (vpratz) [reply bayesflow/simulators/model_comparison_simulator.py:168]: I can imagine this function to be quite cheap to compute, would it make sense to run it completely every time (but only logging the info once)?

**Review by vpratz — COMMENTED** [review]
See my comment on one edge case, I'm not sure if it is a relevant one. What do you think?
Apart from that, the PR looks good to me, I only added minor formatting fixes to the docstring.

File: bayesflow/simulators/model_comparison_simulator.py (hunk 3 above)
Comment (LarsKue) [reply bayesflow/simulators/model_comparison_simulator.py:168]: I agree

File: tests/test_simulators/test_simulators.py [hunk 4]
```diff
@@ -52,8 +53,19 @@ def test_fixed_sample(composite_gaussian, batch_size, fixed_n, fixed_mu):
...
+    if multimodel_key_conflicts.key_conflicts == "drop":
+        samples = multimodel_key_conflicts.sample(batch_size)
         assert set(samples) == {"x", "model_indices"}
-    else:
+    elif multimodel_key_conflicts.key_conflicts == "fill":
+        samples = multimodel_key_conflicts.sample(batch_size)
         assert set(samples) == {"x", "model_indices", "c", "w"}
         assert np.sum(np.isnan(samples["c"])) + np.sum(np.isnan(samples["w"])) == batch_size
+    elif multimodel_key_conflicts.key_conflicts == "error":
+        with pytest.raises(Exception):
```
Comment (LarsKue) [inline tests/test_simulators/test_simulators.py:70]: This is too broad of a check. Use specific exception types.

**Review by LarsKue — CHANGES_REQUESTED** [review]
The changes are looking great. Can we address Valentin's comments before we merge? I also left some minor comments, still.

File: bayesflow/simulators/model_comparison_simulator.py (hunk 3 above)
Comment (Kucharssim) [reply bayesflow/simulators/model_comparison_simulator.py:168]: ok, will do that

**Review by vpratz — APPROVED** [review]
Thanks a lot for the revised changes, they look good to me now.
